### ISSUE TITLE
Class/Type mirrors

### ISSUE THREAD

**Issue by mP1** [mP1/walkingkooka#1767]
MirrorPolicy controls which types, ctor, methods and fields are visible (if any) and invokable.

**Comment by Asmeble** [comment]
to what end?

**Comment by mP1** [comment]
One goal is for a parent or container to use a visitor to rewrite class files replace calls to `Object.getClass()` to a Mirror helper, lets call it JdkReflectionMirror.

This means that all the following refs in the rewritten class file will change to their Mirror equivalents, and basically have the same method name and parameters.

- Class -> MirrorType
- Method -> MirrorMethod
- Field -> MirrorField.

Code like this:

```java
Class k = $something.getClass()
Class k2 = Class.forName("java.io.File")
``` 

becomes

```java
MirrorType k = JdkReflectionMirror.getClass($something).
MirrorType k2 = JdkReflectionMirror.getClass("java.io.File")
```

`Class.getName`, `Class.getSimpleName` are special cases and will continue to work and simply call their equiv methods on the real `Class`.

```
MirrorType {
   String getName() {
      return this.klass.getName();
   }
   String getSimpleName() {
      return this.klass.getSimpleName();
   }
}
```

All other `MirrorType` methods which may then filter the methods that the caller sees etc.

This introduces several advantages, if a container wish to blacklist any file activity, they can fail classes with `java.io.File` references, and any attempts to get a File instance via reflection because my mirrors 
will get in the way. The `SecurityManager` support for fine grained reflection control from what I can tell is very limited and impossible.

If `java.io.File` is blacklisted then  `JdkReflectionMirror.getClass("java.io.File")` in our example will throw a `ClassNotFoundException`.

Nothing is free, and this of course means dynamic languages will be slowed down significantly.

**Comment by Asmeble** [comment]
Have you considered "java Assist" or "java bytebuddy"?

These might get closer to what you meant with some reflection/monkey-punching you're asking for here.

**Comment by mP1** [comment]
Yes. Im probably going to use Object web's ASM to do the class rewriting, its not terribly difficult after all im doing is replacing type refs `java.lang.Class` with `...MirrrorType` and repeat for `java.lang.reflect.Method` and `java.lang.reflect.Field` with their my Mirror equiv.

**Comment by Asmeble** [comment]
Neat and cool

**Comment by mP1** [comment]
But I will look again what new stuff javassist and byte buddy do now as its been a while since ive looked at them.

**Comment by Asmeble** [comment]
Let me know when and if I can help.

I would like to submit a patch.

**Comment by mP1** [comment]
Its just a library, Maven builds, run tests and packages a jar. There arent any releases ATM. Theres no web server etc.

**Comment by mP1** [comment]
If you want to submit patches thats great, all that i ask is that you also write up a few tests to verify stuff works.
